go/vt/vtgate/planbuilder/operators/rewriters.go

```diff
@@ -200,34 +210,60 @@ func bottomUp(
 		return root, NoRewrite
 	}
 
-	oldInputs := root.Inputs()
 	var anythingChanged *ApplyResult
-	newInputs := make([]Operator, len(oldInputs))
-	childID := rootID
-
-	// noLHSTableSet is used to mark which operators that do not send data from the LHS to the RHS
-	// It's only UNION at this moment, but this package can't depend on the actual operators, so
-	// we use this interface to avoid direct dependencies
-	type noLHSTableSet interface{ NoLHSTableSet() }
-
-	for i, operator := range oldInputs {
-		// We merge the table set of all the LHS above the current root so that we can
-		// send it down to the current RHS.
-		// We don't want to send the LHS table set to the RHS if the root is a UNION.
-		// Some operators, like SubQuery, can have multiple child operators on the RHS
-		if _, isUnion := root.(noLHSTableSet); !isUnion && i > 0 {
-			childID = childID.Merge(resolveID(oldInputs[0]))
+
+	switch root := root.(type) {
+	case nullaryOperator:
+		// no inputs, nothing to do
+	case *unaryOperator:
+		newSource, changed := bottomUp(root.Source, rootID, resolveID, rewriter, shouldVisit, false)
+		if DebugOperatorTree && changed.Changed() {
+			fmt.Println(ToTree(newSource))
 		}
-		in, changed := bottomUp(operator, childID, resolveID, rewriter, shouldVisit, false)
+		anythingChanged = anythingChanged.Merge(changed)
+		root.Source = newSource
+	case *binaryOperator:
+		newLHS, changed := bottomUp(root.LHS, rootID, resolveID, rewriter, shouldVisit, false)
 		if DebugOperatorTree && changed.Changed() {
-			fmt.Println(ToTree(in))
+			fmt.Println(ToTree(newLHS))
 		}
 		anythingChanged = anythingChanged.Merge(changed)
-		newInputs[i] = in
-	}
+		root.LHS = newLHS
+		newRHS, changed := bottomUp(root.RHS, rootID, resolveID, rewriter, shouldVisit, false)
+		if DebugOperatorTree && changed.Changed() {
+			fmt.Println(ToTree(newRHS))
+		}
+		anythingChanged = anythingChanged.Merge(changed)
+		root.RHS = newRHS
```

Comment:
nit picking but how about extracting this, so it can be reused for the LHS and RHS

Reply:
I tried, but haven't found a nice way to do it without creating extra intermediate structs, and that's exactly what we are trying to avoid here.

Do you have an example of what you were thinking?